src/version_check.py

```diff
@@ -0,0 +1,244 @@
+    ) -> None:
+        """Constructor for CrossAppVersionChecker object.
+
+        Args:
+            charm: charm to inherit from
+            version: (int), the current version of the desired attribute of the charm
+            relations_to_check: (List), a list of relations who should have compatible versions
+                with the current charm
+            version_validity_range: (Optional Dict), a list of ranges for valid version ranges.
+                If not provided it is assumed that relations on the provided interface must have
+                the same version.
+        """
+        self.charm = charm
+        super().__init__(self.charm, None)
+        # Future PR: upgrade this to a dictionary name versions
+        self.version = version
+        self.relations_to_check = relations_to_check
+
+        for rel in relations_to_check:
+            self.framework.observe(
+                charm.on[rel].relation_created,
+                self.set_version_on_relation_created,
+            )
+
+        # this feature has yet to be implemented, MongoDB does not need it and it is unclear if
+        # this will be extended to other charms. If this code is extended to other charms and
+        # there is a valid usecase we will use the `version_validity_range` variable in the
+        # function `get_invalid_versions`
+        self.version_validity_range = version_validity_range
+
+    def get_invalid_versions(self) -> List[Tuple[str, int]]:
+        """Returns a tuple of (app name, version number), if the version number mismatches.
+
+        Mismatches are decided based on version_validity_range, if version_validity_range is not
+        provided, then the mismatch is expected to match this current app's version number.
+
+        Raises:
+            NoVersionError.
+        """
+        try:
+            invalid_relations = []
+            for relation_name in self.relations_to_check:
+                for relation in self.charm.model.relations[relation_name]:
+                    related_version = relation.data[relation.app][VERSION_CONST]
+                    if int(related_version) != self.version:
+                        invalid_relations.append((relation.app.name, related_version))
+        except KeyError:
+            raise NoVersionError(f"Expected {relation.app.name} to have version info.")
+
+        return invalid_relations
+
+    def get_version_of_related_app(self, related_app_name: str) -> int:
+        """Returns a int for the version of the related app.
+
+        Raises:
+            NoVersionError.
+        """
+        try:
+            for relation_name in self.relations_to_check:
+                for rel in self.charm.model.relations[relation_name]:
+                    if rel.app.name == related_app_name:
+                        return int(rel.data[rel.app][VERSION_CONST])
+        except KeyError:
+            pass
+
+        raise NoVersionError(f"Expected {related_app_name} to have version info.")
```

Comment:
This exception is not caught in `charm.py`. We could probably return None here instead of raising

Reply:
Two things:
> This exception is not caught in charm.py

This is on purpose. There is no need to check for this exception in `src/charm.py` at that point in time. This is because it would have caused an error on an earlier line. 

>  We could probably return None here instead of raising

I would prefer to keep it as is. I want to be consistent with the error of `NoVersionError` occurring when there is no version information present.